File: 061__Rotate_List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if not (head and head.next):
            return head

        tail = head
        length = 1
        while tail.next:  # get length and find the tail
            tail = tail.next
            length += 1
        tail.next = head  # link it to head, just like a ring

        # ring: [back k step] == [forward (-k mod length) step]
        step = (-k) % length
        while step:
            tail = tail.next  # just step forward
            step -= 1

        head = tail.next
        tail.next = None
        return head

File: Java/061__Rotate_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || k == 0) {
            return head;
        }
        int length = 1;
        ListNode tail = head;
        while (tail.next != null) {
            tail = tail.next;
            length++;
        }
        k = k % length;

        tail = head;
        for (int i = 0; i < k; i++) {
            tail = tail.next != null ? tail.next : head;
        }
        // k == size of list, to be head finally.
        if (tail == head) {
            return head;
        }

        ListNode prev = head;
        while (tail.next != null) {
            tail = tail.next;
            prev = prev.next;
        }
        ListNode newHead = prev.next;
        prev.next = null;
        tail.next = head;
        return newHead;
    }
}
